README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> HashiCorp Vault based Credential Store and Discovery Service for Node.js / ES2017

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The Components module contains standard component definitions that can be used to build applications and services.

The module contains the following packages:
- **Auth** - authentication credential stores
- **Build** - basic factories for constructing objects

<a name="links"></a> Quick links:

* [Logging](https://www.pipservices.org/recipies/logging)
* [Configuration](https://www.pipservices.org/recipies/configuration) 
* [API Reference](https://pip-services3-nodex.github.io/pip-services3-components-nodex/globals.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use

Install the NPM package as
```bash
npm install pip-services3-components-nodex --save
```

Example how to use Logging and Performance counters.
Here we are going to use CompositeLogger and CompositeCounters components.
They will pass through calls to loggers and counters that are set in references.

```typescript
import { ConfigParams } from 'pip-services3-commons-nodex'; 
import { IConfigurable } from 'pip-services3-commons-nodex'; 
import { IReferences } from 'pip-services3-commons-nodex'; 
import { IReferenceable } from 'pip-services3-commons-nodex'; 
import { CompositeLogger } from 'pip-services3-components-nodex'; 
import { CompositeCounters } from 'pip-services3-components-nodex'; 

export class MyComponent implements IConfigurable, IReferenceable {
  private _logger: CompositeLogger = new CompositeLogger();
  private _counters: CompositeCounters = new CompositeCounters();
  
  public configure(config: ConfigParams): void {
    this._logger.configure(config);
  }
  
  public setReferences(refs: IReferences): void {
    this._logger.setReferences(refs);
    this._counters.setReferences(refs);
  }
  
  public async myMethod(correlationId: string, param1: any): Promise<void> {
    this._logger.trace(correlationId, "Executed method mycomponent.mymethod");
    this._counters.increment("mycomponent.mymethod.exec_count", 1);
    let timing = this._counters.beginTiming("mycomponent.mymethod.exec_time");
    try {
      ....
    } finally {
      timing.endTiming();
    } catch (err) {
      if (err) {
        this._logger.error(correlationId, err, "Failed to execute mycomponent.mymethod");
        this._counters.increment("mycomponent.mymethod.error_count", 1);
      }
    }
    ...
  }
}
```

Example how to get connection parameters and credentials using resolvers.
The resolvers support "discovery_key" and "store_key" configuration parameters
to retrieve configuration from discovery services and credential stores respectively.

```typescript
import { ConfigParams } from 'pip-services3-commons-nodex'; 
import { IConfigurable } from 'pip-services3-commons-nodex'; 
import { IReferences } from 'pip-services3-commons-nodex'; 
import { IReferenceable } from 'pip-services3-commons-nodex'; 
import { IOpenable } from 'pip-services3-commons-nodex'; 
import { ConnectionParams } from 'pip-services3-components-nodex'; 
import { ConnectionResolver } from 'pip-services3-components-nodex'; 
import { CredentialParams } from 'pip-services3-components-nodex'; 
import { CredentialResolver } from 'pip-services3-components-nodex'; 

export class MyComponent implements IConfigurable, IReferenceable, IOpenable {
  private _connectionResolver: ConnectionResolver = new ConnectionResolver();
  private _credentialResolver: CredentialResolver = new CredentialResolver();
  
  public configure(config: ConfigParams): void {
    this._connectionResolver.configure(config);
    this._credentialResolver.configure(config);
  }
  
  public setReferences(refs: IReferences): void {
    this._connectionResolver.setReferences(refs);
    this._credentialResolver.setReferences(refs);
  }
  
  ...
  
  public async open(correlationId: string): Promise<void> {
    let connection = await this._connectionResolver.resolve(correlationId);
    let credential = await this._credentialResolver.lookup(correlationId);

    let host = connection.getHost();
    let port = connection.getPort();
    let user = credential.getUsername();
    let pass = credential.getPassword();
  }
}

// Using the component
let myComponent = new MyComponent();

myComponent.configure(ConfigParams.fromTuples(
  'connection.host', 'localhost',
  'connection.port', 1234,
  'credential.username', 'anonymous',
  'credential.password', 'pass123'
));

await myComponent.open(null);
```

Example how to use caching and locking.
Here we assume that references are passed externally.

```typescript
import { Descriptor } from 'pip-services3-commons-nodex'; 
import { References } from 'pip-services3-commons-nodex'; 
import { IReferences } from 'pip-services3-commons-nodex'; 
import { IReferenceable } from 'pip-services3-commons-nodex'; 
import { ILock } from 'pip-services3-components-nodex'; 
import { MemoryLock } from 'pip-services3-components-nodex'; 
import { ICache } from 'pip-services3-components-nodex'; 
import { MemoryCache } from 'pip-services3-components-nodex'; 

export class MyComponent implements IReferenceable {
  private _cache: ICache;
  private _lock: ILock;
  
  public setReferences(refs: IReferences): void {
    this._cache = refs.getOneRequired<ICache>(new Descriptor("*", "cache", "*", "*", "1.0"));
    this._lock = refs.getOneRequired<ILock>(new Descriptor("*", "lock", "*", "*", "1.0"));
  }
  
  public async myMethod(correlationId: string, param1: any): Promise<any> {
    // First check cache for result
    result := await this._cache.retrieve(correlationId, "mykey");
    if (result != null) {
      return result;
    }
      
    // Lock..
    await this._lock.acquireLock(correlationId, "mykey", 1000, 1000);
    
    // Do processing
    ...
    
    // Store result to cache async
    this._cache.store(correlationId, "mykey", result, 3600000);

    // Release lock async
    this._lock.releaseLock(correlationId, "mykey");

    return result;
  }
}

// Use the component
let myComponent = new MyComponent();

myComponent.setReferences(References.fromTuples(
  new Descriptor("pip-services", "cache", "memory", "default", "1.0"), new MemoryCache(),
  new Descriptor("pip-services", "lock", "memory", "default", "1.0"), new MemoryLock(),
);

result := await myComponent.myMethod(null);
```

If you need to create components using their locators (descriptors) implement 
component factories similar to the example below.

```typescript
import { Factory } from 'pip-services3-components-nodex';
import { Descriptor } from 'pip-services3-commons-nodex';

export class MyFactory extends Factory {
  public static myComponentDescriptor: Descriptor = new Descriptor("myservice", "mycomponent", "default", "*", "1.0");
  
  public MyFactory() {
    super();
    
    this.registerAsType(MyFactory.myComponentDescriptor, MyComponent);    
  }
}

// Using the factory

let myFactory = MyFactory();

let myComponent1 = myFactory.create(new Descriptor("myservice", "mycomponent", "default", "myComponent1", "1.0");
let myComponent2 = myFactory.create(new Descriptor("myservice", "mycomponent", "default", "myComponent2", "1.0");

...
```

## Develop

For development you shall install the following prerequisites:
* Node.js 8+
* Visual Studio Code or another IDE of your choice
* Docker
* Typescript

Install dependencies:
```bash
npm install
```

Compile the code:
```bash
tsc
```

Run automated tests:
```bash
npm test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by **Sergey Seroukhov**.

The documentation is written by **Mark Makarychev**.


CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Vault based Credential Store and Discovery Service for Node.js / ES2017 Changelog

## <a name="1.0.1-1.0.2"></a> 1.0.1-1.0.2 (2023-05-24) 


### Features
- Added rewrite flag for *loadVaultCredentials* and *loadVaultConnections*

### Bug fixes
- Fixed Vault result prepearing

## <a name="1.0.0"></a> 1.0.0 (2023-05-23) 

### Features
* Implement Credential Store and Discovery Service for HashiCorp Vault



test/connect/VaultDiscovery.test.ts
const assert = require('chai').assert;

import { ConfigParams } from 'pip-services3-commons-nodex';
import { ConnectionParams } from 'pip-services3-components-nodex';

import { VaultDiscovery } from '../../src/connect/VaultDiscovery';


suite('VaultDiscovery', ()=> {
    let discovery: VaultDiscovery

    let host = process.env['VAULT_HOST'] ?? 'localhost';
    let port = process.env['VAULT_PORT'] ?? '8201';
    
    let vaultConfig = ConfigParams.fromTuples(
        'connection.protocol', 'http',
        "connection.host", host,
        "connection.port", port,
        'credential.auth_type', 'userpass',
        'credential.username', 'user',
        'credential.password', 'pass',
    );  

    let connections = ConfigParams.fromTuples(
        "connKey1.host", "10.1.1.100",
        "connKey1.port", "8080",
        "connKey2.host", "10.1.1.101",
        "connKey2.port", "8082",
    );

    setup(async () => {
        discovery = new VaultDiscovery();
        discovery.configure(vaultConfig);
        await discovery.open(null);
        await discovery.loadVaultCredentials(connections, true);
    });

    test('Resolve connections', async () => {
        // Resolve one
        let connection = await discovery.resolveOne("123", "connKey1");

        assert.equal("10.1.1.100", connection.getHost());
        assert.equal(8080, connection.getPort());        

        connection = await discovery.resolveOne("123", "connKey2");

        assert.equal("10.1.1.101", connection.getHost());
        assert.equal(8082, connection.getPort());      

        // Resolve all
        await discovery.register(null, "connKey1",
            ConnectionParams.fromTuples("host", "10.3.3.151")
        );

        let connections = await discovery.resolveAll("123", "connKey1");
        
        assert.isTrue(connections.length > 1);

        await discovery.register(null, "connKey2",
            ConnectionParams.fromTuples("host", "15.7.7.1", "port", "8053")
        );

        connections = await discovery.resolveAll("123", "connKey2");

        assert.isTrue(connections.length > 1);

        await discovery.register(null, "connKey3",
            ConnectionParams.fromTuples("host", "7.7.0.0", "port", "8585")
        );

        connections = await discovery.resolveAll("123", "connKey3");

        assert.isTrue(connections.length == 1);

    });

});

test/auth/VaultCredentialStore.test.ts
const assert = require('chai').assert;

import { ConfigParams } from 'pip-services3-commons-nodex';
import { CredentialParams } from 'pip-services3-components-nodex';

import { VaultCredentialStore } from '../../src/auth/VaultCredentialStore';


suite('VaultCredentialStore', ()=> {
    let credentialStore: VaultCredentialStore
    
    let host = process.env['VAULT_HOST'] ?? 'localhost';
    let port = process.env['VAULT_PORT'] ?? '8201';

    let vaultConfig = ConfigParams.fromTuples(
        'connection.protocol', 'http',
        "connection.host", host,
        "connection.port", port,
        'credential.auth_type', 'userpass',
        'credential.username', 'user',
        'credential.password', 'pass',
    );

    let credentials = ConfigParams.fromTuples(
        'credKey1.user', 'user1',
        'credKey1.pass', 'pass1',
        'credKey2.user', 'user2',
        'credKey2.pass', 'pass2',
    );

    setup(async () => {
        credentialStore = new VaultCredentialStore();
        credentialStore.configure(vaultConfig);
        await credentialStore.open(null);
        await credentialStore.loadVaultCredentials(credentials, true);
    });

    test('Lookup and store test', async () => {
        let cred1 = await credentialStore.lookup('123', 'credKey1');
        let cred2 = await credentialStore.lookup('123', 'credKey2');

        assert.equal(cred1.getUsername(), "user1");
        assert.equal(cred1.getPassword(), "pass1");
        assert.equal(cred2.getUsername(), "user2");
        assert.equal(cred2.getPassword(), "pass2");

        let credConfig = CredentialParams.fromTuples(
            'user', 'user3',
            'pass', 'pass3',
            'access_id', '123'
        );

        await credentialStore.store(null, 'credKey3', credConfig);

        let cred3 = await credentialStore.lookup('123', 'credKey3');

        assert.equal(cred3.getUsername(), "user3");
        assert.equal(cred3.getPassword(), "pass3");
        assert.equal(cred3.getAccessId(), "123");

        credConfig = CredentialParams.fromTuples(
            'user', 'new_user',
            'pass', 'new_pass',
            'access_id', 'new_id'
        );

        await credentialStore.store(null, 'credKey1', credConfig);

        cred1 = await credentialStore.lookup('123', 'credKey1');

        assert.equal(cred1.getUsername(), "new_user");
        assert.equal(cred1.getPassword(), "new_pass");
        assert.equal(cred1.getAccessId(), "new_id");
    });

});

src/connect/VaultDiscovery.ts
/** @module connect */

import { ApplicationException, ConfigException, ConfigParams, ConnectionException, IConfigurable, IOpenable, IReferenceable, IReferences } from 'pip-services3-commons-nodex';
import { IReconfigurable } from 'pip-services3-commons-nodex';

import { CompositeLogger, ConnectionParams, ConnectionResolver, CredentialParams, CredentialResolver, MemoryDiscovery } from 'pip-services3-components-nodex';
import { IDiscovery } from 'pip-services3-components-nodex';

/**
 * Discovery service that keeps connections in memory.
 * 
 * ### Configuration parameters ###
 * 
 * - connection(s):           
 *   - discovery_key:         (optional) a key to retrieve the connection from [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]]
 *   - host:                  host name or IP address
 *   - port:                  port number
 *   - uri:                   resource URI or connection string with all parameters in it
 *   - proxy_enable:          enable proxy (default false)
 *   - proxy_host:            proxy host name
 *   - proxy_port:            proxy port number
 * - credential(s):
 *   - store_key:             key to retrieve parameters from credential store
 *   - username:              set user name for ldap and userpass auth type, role_id for approle and k8s auth type, cert_name for cert auth type
 *   - password:              user password for ldap and userpass auth type, secret_id for approle auth type, token for k8s and cert_name auth type
 *   - auth_type:             auth type (approle, ldap, userpass, k8s, cert) default - userpass
 *   - file_cert:             client certificate file for https mode
 *   - file_key:              client key file for https mode
 *   - file_cacert:           root CA cert path for https mode
 * - options:
 *   - root_path:             root path after the base URL
 *   - timeout:               default timeout in milliseconds (default: 5 sec)
 *   - namespace:             namespace (multi-tenancy) feature available on all Vault Enterprise versions   
 * @see [[IDiscovery]]
 * @see [[ConnectionParams]]
 * 
 *     
 *     let discovery = new VaultDiscovery();
 *     discovery.open();
 *     
 *     let connection = await discovery.resolve("123", "key1");
 *     // Result: host=10.1.1.100;port=8080
 *     
 */
export class VaultDiscovery implements IDiscovery, IReconfigurable, IReferenceable, IConfigurable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();
    private _credentialResolver: CredentialResolver = new CredentialResolver();

    //connection params
    private _proxy_enable: boolean = false;
    private _proxy_port: number;
    private _proxy_host: string;

    // credentials
    private _auth_type: string = "userpass";
    private _file_cert: string;
    private _file_key: string;
    private _file_cacert: string;

    // options
    private _timeout: number = 5000;
    private _root_path: string;
    private _namespace: string;

    private _client: any = null;
    private _token: string;

    /** 
     * The logger.
     */
    protected _logger: CompositeLogger = new CompositeLogger();

    /**
    * Configures component by passing configuration parameters.
    * 
    * @param config    configuration parameters to be set.
    */
    public configure(config: ConfigParams): void {
        this._connectionResolver.configure(config);
        this._credentialResolver.configure(config);

        this._logger.configure(config);
        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._namespace = config.getAsStringWithDefault('options.namespace', this._namespace);
    }

    /**
    * Reads connections from configuration parameters.
    * And save it to Vault.
    * 
    * @param config   configuration parameters to be read
    * @param rewrite   rewrite flag if key exists
    */
    public async loadVaultCredentials(config: ConfigParams, rewrite?: boolean): Promise<void> {
        let items: Map<string, ConnectionParams[]> = new Map();

        if (config.length() > 0) {
            let connectionSections: string[] = config.getSectionNames();
            for (let index = 0; index < connectionSections.length; index++) {
                let key = connectionSections[index]
                let value: ConfigParams = config.getSection(key);

                let connectionsList = items.get(key) ?? [];
                connectionsList.push(new ConnectionParams(value));
                items.set(key, connectionsList);
            }
        }

        // Register all connections in vault
        for (let key of items.keys()) {
            for (let conn of items.get(key)) {
                if (!rewrite) {
                    try {
                        await this._client.readKVSecret(this._token, key);
                    } catch (ex) {
                        if (ex.response && ex.response.status == 404) {
                            await this.register(null, key, conn);
                        } else {
                            throw ex;
                        }
                    }
                } else {
                    await this.register(null, key, conn);
                }
            }
        }
    }

    /**
    * Sets references to dependent components.
    * 
    * @param references 	references to locate the component dependencies. 
    */
    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
        this._credentialResolver.setReferences(references);
        this._logger.setReferences(references);
    }
    /**
     * Checks if the component is opened.
     * 
     * @returns true if the component has been opened and false otherwise.
    */
    public isOpen(): boolean {
        return this._client != null;
    }

    /**
     *  Helper method for resolve all additonal parameters
     */
    private resolveConfig(correlationId: string, connection: ConnectionParams, credential: CredentialParams) {

        // check configuration
        if (connection == null) {
            throw new ConfigException(
                correlationId,
                "NO_CONNECTION",
                "Connection is not configured"
            );
        }

        if (credential == null) {
            throw new ConfigException(
                correlationId,
                "NO_CREDENTIAL",
                "Credentials is not configured"
            );
        }

        // resolve additional credential params
        this._auth_type = credential.getAsStringWithDefault("auth_type", "userpass");
        this._file_cert = credential.getAsNullableString("file_cert");
        this._file_key = credential.getAsNullableString("file_key");
        this._file_cacert = credential.getAsNullableString("file_cacert");

        // resolve additionla connection params
        this._proxy_enable = connection.getAsBooleanWithDefault("proxy_enable", false);
        this._proxy_port = connection.getAsNullableInteger("proxy_port");
        this._proxy_host = connection.getAsNullableString("proxy_host");
    }

    /**
     *  Helper method for compose uri
     */
    private composeUri(correlationId: string, connection: ConnectionParams): string {

        if (connection.getUri() != null) {
            let uri = connection.getUri();
            if (uri) return uri;
        }

        let host = connection.getHost();
        if (host == null) {
            throw new ConfigException(
                correlationId,
                "NO_HOST",
                "Connection host is not set"
            );
        }

        let port = connection.getPort();
        if (port == 0) {
            throw new ConfigException(
                correlationId,
                "NO_PORT",
                "Connection port is not set"
            );
        }

        let protocol = connection.getProtocol();
        if (protocol == null) {
            throw new ConfigException(
                correlationId,
                "NO_PROTOCOL",
                "Connection protocol is not set"
            );
        }

        return protocol + '://' + host + ':' + port + '/v1';
    }

    /**
     * Opens the component.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async open(correlationId: string): Promise<void> {

        let connection = await this._connectionResolver.resolve(correlationId);
        let credential = await this._credentialResolver.lookup(correlationId);
        this.resolveConfig(correlationId, connection, credential);

        let options: any =
        {
            https: connection.getProtocol() === "https",
            baseUrl: this.composeUri(correlationId, connection),
            timeout: this._timeout,
            proxy: false,
        };

        // configure additional options
        if (this._root_path != null) {
            options.rootPath = this._root_path;
        }

        if (this._namespace != null) {
            options.namespace = this._namespace;
        }

        // configure https connection
        if (connection.getProtocol() === "https") {
            options.cert = this._file_cert;
            options.key = this._file_key;
            options.cacert = this._file_cacert;
        }

        // configure proxy
        if (this._proxy_enable) {
            options.proxy = {
                host: this._proxy_host,
                port: this._proxy_port
            }
        }

        // configure credentials
        let username: string;
        let password: string;

        if (credential != null) {
            username = credential.getUsername();
            password = credential.getPassword();
        }

        const Vault = require('hashi-vault-js');
        this._client = new Vault(options);
        const status = await this._client.healthCheck();

        // resolve status
        if (status.isVaultError || status.response) {
            let err = new ApplicationException("ERROR", correlationId, "OPEN_ERROR", status.vaultHelpMessage);
            this._logger.error(correlationId, err, status.vaultHelpMessage, status.response)
            this._client = null;
            throw err;
        } else if (status.sealed) {
            let err = new ApplicationException("ERROR", correlationId, "OPEN_ERROR", "Vault server is sealed!")
            this._logger.error(correlationId, err, "Vault server is sealed!")
            this._client = null;
            throw err // TODO: Decide, does need to throw error?
        }

        this._logger.debug(correlationId, "Vault status:", status)

        // open connection and get API token
        try {
            switch (this._auth_type) {
                case "approle": {
                    this._token = (await this._client.loginWithAppRole(username, password)).client_token;
                    break;
                }
                case "ldap": {
                    this._token = (await this._client.loginWithLdap(username, password)).client_token;
                    break;
                }
                case "userpass": {
                    this._token = (await this._client.loginWithUserpass(username, password)).client_token;
                    break;
                }
                case "k8s": {
                    this._token = (await this._client.loginWithK8s(username, password)).client_token;
                    break;
                }
                case "cert": {
                    this._token = (await this._client.loginWithCert(username, password)).client_token;
                    break;
                }
                default: {
                    this._token = (await this._client.loginWithUserpass(username, password)).client_token;
                    break;
                }
            }
        } catch (ex) {
            let err = new ConnectionException(correlationId, "LOGIN_ERROR", "Can't login to Vault server").withCause(ex);
            this._logger.error(correlationId, ex, "Can't login to Vault server")
            this._client = null;
            throw err
        }
        this._logger.info(correlationId, "Vault Discovery Service opened with %s auth mode", this._auth_type);
        return
    }

    /**
    * Closes component and frees used resources.
    * 
    * @param correlationId 	(optional) transaction id to trace execution through call chain.
    */
    public async close(correlationId: string): Promise<void> {
        if (this.isOpen()) {
            this._client = null;
        }
        this._logger.info(correlationId, "Vault Discovery Service closed");
    }

    /**
     * Registers connection parameters into the discovery service.
     *
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a key to uniquely identify the connection parameters.
     * @param credential        a connection to be registered.
     * @returns 			    the registered connection parameters.
     */
    public async register(correlationId: string, key: string, connection: ConnectionParams): Promise<ConnectionParams> {
        if (this.isOpen()) {
            try {
                let connections = [];
                let version = 0;
                try {
                    let res = await this._client.readKVSecret(this._token, key);
                    version = res.metadata.version;
                    // Check if connection already exists
                    if (res.data && res.data.connections) {
                        for (let conn of res.data.connections) {
                            if (connection.getHost() == conn.host && connection.getPort() == (conn.port ?? 0)) {
                                this._logger.info(correlationId, 'Connection already exists via key ' + key + ': ' + connection);
                                return connection;
                            }
                        }
                    
                        for (let conn of res.data.connections)
                            connections.push(new ConnectionParams(conn).getAsObject());
                    }
                } catch (ex) {
                    if (ex.response && ex.response.status == 404) {
                        // pass
                    } else {
                        throw ex;
                    }
                }

                connections.push(connection.getAsObject());
                
                if (version > 0) {
                    await this._client.updateKVSecret(this._token, key, { connections: connections }, version);
                } else {
                    await this._client.createKVSecret(this._token, key, { connections: connections });
                }
                
                this._logger.debug(correlationId, 'Register via key ' + key + ': ' + connection);
                return connection;
            } catch (ex) {
                this._logger.error(correlationId, ex, "Can't store KV to Vault with key: " + key);
            }

        }
    }

    /**
     * Resolves a single connection parameters by its key.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a key to uniquely identify the connection.
     * @returns                 a found connection parameters or <code>null</code> otherwise
     */
    public async resolveOne(correlationId: string, key: string): Promise<ConnectionParams> {
        if (this.isOpen()) {
            try {
                let res = await this._client.readKVSecret(this._token, key);
                this._logger.debug(correlationId, 'Resolved connection for ' + key + ': ', res);
                let connection: ConnectionParams;
                if (res.data && res.data.connections && res.data.connections.length > 0)
                    connection = new ConnectionParams(res.data.connections[0]);
                return connection;
            } catch (ex) {
                this._logger.error(correlationId, ex, "Can't resolve KV from Vault with key: " + key);
            }

        }
    }

    /**
     * Resolves all connection parameters by their key.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a key to uniquely identify the connections.
     * @returns                 all found connection parameters
     */
    public async resolveAll(correlationId: string, key: string): Promise<ConnectionParams[]> {
        if (this.isOpen()) {
            try {
                let res = await this._client.readKVSecret(this._token, key)
                this._logger.debug(correlationId, 'Resolved connections for ' + key + ': ', res);

                let connections: ConnectionParams[] = [];

                for (let conn of res.data.connections)
                    connections.push(new ConnectionParams(conn));
                
                return connections;
            } catch (ex) {
                this._logger.error(correlationId, ex, "Can't resolve KV from Vault with key: " + key);
            }
        }

    }
}